src/HAL/Hal.c

```diff
@@ -0,0 +1,314 @@
+    pioConfigTx->pinsConfig.control = Pio_Control_PeripheralC;
+}
+
+inline static void
+Hal_uart_init_pmc(Uart_Id id)
+{
+    switch(id) {
+        case Uart_Id_0:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioA);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart0);
+            break;
+        case Uart_Id_1:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioA);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart1);
+            break;
+        case Uart_Id_2:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioD);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart2);
+            break;
+        case Uart_Id_3:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioD);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart3);
+            break;
+        case Uart_Id_4:
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_PioD);
+            Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart4);
+            break;
+    }
+}
+
+inline inline static void
+Hal_uart_init_nvic(Uart_Id id)
+{
+    switch(id) {
+        case Uart_Id_0:
+            Nvic_enableInterrupt(Nvic_Irq_Uart0);
+            Nvic_setInterruptPriority(Nvic_Irq_Uart0, 5);
+            break;
+        case Uart_Id_1:
+            Nvic_enableInterrupt(Nvic_Irq_Uart1);
+            Nvic_setInterruptPriority(Nvic_Irq_Uart1, 5);
+            break;
+        case Uart_Id_2:
+            Nvic_enableInterrupt(Nvic_Irq_Uart2);
+            Nvic_setInterruptPriority(Nvic_Irq_Uart2, 5);
+            ;
+            break;
+        case Uart_Id_3:
+            Nvic_enableInterrupt(Nvic_Irq_Uart3);
+            Nvic_setInterruptPriority(Nvic_Irq_Uart3, 5);
+            break;
+        case Uart_Id_4:
+            Nvic_enableInterrupt(Nvic_Irq_Uart4);
+            Nvic_setInterruptPriority(Nvic_Irq_Uart4, 5);
+            break;
+    }
+}
+
+void
+Hal_console_usart_init(void)
+{
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* CCFG_SYSIO = (uint32_t*)0x40088114;
+
+    // Assign the PB4 pin to the PIO controller (TDI is the default function)
+    *CCFG_SYSIO |= 1u << 4u;
+
+    Pmc_enablePeripheralClk(Pmc_PeripheralId_PioB);
+    Pmc_enablePeripheralClk(Pmc_PeripheralId_Usart1);
+
+    Pio pioB;
+    Pio_init(Pio_Port_B, &pioB);
+
+    Pio_Pin_Config pinConf;
+    pinConf.control = Pio_Control_PeripheralD;
+    pinConf.direction = Pio_Direction_Input;
+    pinConf.pull = Pio_Pull_None;
+    pinConf.filter = Pio_Filter_None;
+    pinConf.isMultiDriveEnabled = false;
+    pinConf.isSchmittTriggerDisabled = false;
+    pinConf.irq = Pio_Irq_None;
+
+    Pio_setPinsConfig(&pioB, PIO_PIN_4, &pinConf);
+
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* US_CR = (uint32_t*)0x40028000;
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* US_MR = (uint32_t*)0x40028004;
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* US_BRGR = (uint32_t*)0x40028020;
+
+    *US_MR = 0;         // Clear
+    *US_MR |= 0u << 0;  // USART_MODE NORMAL
```

Comment:
it would be nice if these were named numbers, possibly defined in Usart_Registers.h